forwardproxy.go

```diff
@@ -28,152 +14,389 @@ import (
 	"net"
 	"net/http"
 	"net/url"
+	"os"
 	"strconv"
 	"strings"
 	"sync"
 	"time"
 
-	"github.com/caddyserver/caddy/caddyhttp/httpserver"
+	"github.com/caddyserver/caddy/v2"
+	"github.com/caddyserver/caddy/v2/caddyconfig/caddyfile"
+	"github.com/caddyserver/caddy/v2/caddyconfig/httpcaddyfile"
+	"github.com/caddyserver/caddy/v2/modules/caddyhttp"
 	"github.com/caddyserver/forwardproxy/httpclient"
+	"golang.org/x/net/proxy"
+	"go.uber.org/zap"
 )
 
-type ForwardProxy struct {
-	Next httpserver.Handler
+func init() {
+	caddy.RegisterModule(Handler{})
+	httpcaddyfile.RegisterHandlerDirective("forward_proxy", parseCaddyfile)
+}
 
-	authRequired    bool
-	authCredentials [][]byte // slice with base64-encoded credentials
+type ProbeResistance struct {
+	Domain string `json:"domain,omitempty"`
+}
 
-	hideIP  bool
-	hideVia bool
+type Handler struct {
+	logger *zap.Logger
 
-	pacFilePath string
+	PACPath string `json:"pac_path,omitempty"`
 
-	hostname string // do not intercept requests to the hostname (except for hidden link)
-	port     string // port on which chain with forwardproxy is listening on
+	HideIP  bool `json:"hide_ip,omitempty"`
+	HideVia bool `json:"hide_via,omitempty"`
 
-	probeResistDomain  string
-	probeResistEnabled bool
+	// // host and port of the proxy; i.e. when configuring a client,
+	// // you give it the host and port of the proxy to use
+	// hostname string // do not intercept requests to the hostname (except for hidden link)
+	// port     string // port on which chain with forwardproxy is listening on
+	Hosts caddyhttp.MatchHost `json:"hosts,omitempty"`
 
-	dialTimeout time.Duration // for initial tcp connection
+	ProbeResistance *ProbeResistance `json:"probe_resistance,omitempty"`
 
-	httpTransport http.Transport
+	DialTimeout caddy.Duration `json:"dial_timeout,omitempty"` // for initial tcp connection
 
+	httpTransport *http.Transport
+
+	Upstream string `json:"upstream,omitempty"`
 	// overridden dialContext allow to redirect requests to upstream proxy
 	dialContext func(ctx context.Context, network, address string) (net.Conn, error)
 	upstream    *url.URL // address of upstream proxy
 
-	aclRules         []aclRule
-	whitelistedPorts []int
+	ACL              []ACLRule `json:"acl,omitempty"`
+	WhitelistedPorts []int     `json:"whitelisted_ports,omitempty"`
+
+	aclRules []aclRule
+
+	// TODO: temporary - we should try to reuse existing authentication modules instead!
+	BasicauthUser string `json:"auth_user,omitempty"`
+	BasicauthPass string `json:"auth_pass,omitempty"`
```

Comment:
Why do we add those 2 fields, instead of using the existing authCredentials slice? How does this temporary solution help tests and why can't we re-use existing credentials checking code?